### PR TITLE
Freeze version to avoid dup

### PR DESCRIPTION
**Pull request by mvz**
Avoiding the creation of a new string speeds up `Gem::Version#<=>`.

Benchmark:

```ruby
require 'benchmark'
n = 500000

orig = Gem::Version.new('1.4-b-4')
same = Gem::Version.new('1.4-b-4')
other = Gem::Version.new('1.4-b-5')
Benchmark.bm(7) do |x|
  x.report('self') { n.times { orig <=> same } }
  x.report('other') { n.times { orig <=> other } }
end
```

Original result:
```
              user     system      total        real
self      0.270000   0.000000   0.270000 (  0.271262)
other     0.880000   0.000000   0.880000 (  0.886234)
```

With this change applied:
```
              user     system      total        real
self      0.120000   0.000000   0.120000 (  0.111093)
other     0.680000   0.000000   0.680000 (  0.688053)
```

### CODE REVIEW

**Comment by drbrain** [comment]
This changes the API of Version so I can't commit it for 2.x as-is.

Instead can we leave #version and #to_s alone and add a new accessor for the frozen version that #<=> calls?

**Comment by mvz** [comment]
Good point. I added a protected method for this.

**Comment by tenderlove** [comment]
I'm super confused.  How does this avoid string creation?

**Comment by mvz** [comment]
It avoids string creation in the `#<=>` method by not calling `.dup` there.

**Comment by mvz** [comment]
As in, `other.version` would call `.dup`.

**Comment by copiousfreetime** [comment]
@mvz would you update this pull request against master and we can get this resolved.

**Comment by mvz** [comment]
@copiousfreetime It looks like @tenderlove implemented essentially the same fix separately in 9fd7533e, by creating a protected method called `_version`. This is `version_string` in this PR. This makes the `freeze` the only real change here, and an implementation detail that probably doesn't make a difference.

TL;DR: I don't think this PR adds anything anymore. WDYT?

**Comment by indirect** [comment]
Thanks for checking into it.
